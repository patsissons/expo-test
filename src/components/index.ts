export * from './state';
export * from './actions';
export * from './context';
export * from './AppComponent';
export * from './startup/Startup';
