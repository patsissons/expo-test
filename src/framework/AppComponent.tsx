import React from 'react';
import { NavigationScreenProps } from 'react-navigation';
import { ReduxProvider } from './context';
import * as styles from './styles';
import * as renderHelpers from './renderHelpers';
import { ContextState } from './state';

export namespace AppComponent {
  export interface Props<TState> extends ReduxProvider.ComponentProps<TState> {
  }

  export interface ContextProps {
    context: ContextState;
  }

  export interface ContextNavigationScreenProps<P = {}, O = {}> extends NavigationScreenProps<P, O> {
    screenProps: ContextProps;
  }
}

export abstract class AppComponent<P extends AppComponent.Props<TState>, TState, S = {}, SS = any> extends React.Component<P, S, SS> {
  constructor(props: P) {
    super(props);

    this.state = {} as S;
  }

  protected get L() {
    return this.props.context.localization;
  }

  protected get R() {
    return renderHelpers;
  }

  protected get S() {
    return styles;
  }
}
