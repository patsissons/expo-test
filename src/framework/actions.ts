import { Action } from 'redux';
import { NavigationParams } from 'react-navigation';
import { createTypeNameMap, isString } from '../utils';

export type Actions = Readonly<typeof actions>;

export interface RequestAction<T = {}> extends Action {
  payload?: T;
}

export interface LocalePayload {
  locale: string;
}

export interface DemoPayload {
  amount: number;
}

export function isLocalePayload(payload: any): payload is LocalePayload {
  const localePayload: LocalePayload = payload;

  return (
    localePayload != null &&
    isString(localePayload.locale)
  );
}

export function getRequestAction(type: string): RequestAction;
export function getRequestAction<T>(type: string, payload: T): RequestAction<T>;
export function getRequestAction<T>(type: string, payload?: T): RequestAction<T> {
  return {
    type,
    payload,
  };
}

enum Context {
  LoadEnv,
  LoadLocale,
  LoadLocalization,
}

enum Navigation {
  NavigationBack,
  NavigationToMain,
  NavigationToLogin,
}

enum Demo {
  DemoIncrement,
  DemoDecrement,
}

export const actionTypes = {
  context: createTypeNameMap(Context),
  navigation: createTypeNameMap(Navigation),
  demo: createTypeNameMap(Demo),
};

export const actions = {
  context: {
    loadEnv: () => getRequestAction(actionTypes.context.LoadEnv),
    loadLocale: () => getRequestAction(actionTypes.context.LoadLocale),
  },
  navigation: {
    goBack: (params?: NavigationParams) => getRequestAction(actionTypes.navigation.NavigationBack, params),
    toMain: (params?: NavigationParams) => getRequestAction(actionTypes.navigation.NavigationToMain, params),
    toLogin: (params?: NavigationParams) => getRequestAction(actionTypes.navigation.NavigationToLogin, params),
  },
  demo: {
    increment: (payload?: DemoPayload) => getRequestAction(actionTypes.demo.DemoIncrement, payload),
    decrement: (payload?: DemoPayload) => getRequestAction(actionTypes.demo.DemoDecrement, payload),
  },
};
