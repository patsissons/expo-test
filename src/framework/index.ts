export * from './state';
export * from './actions';
export * from './context';
export * from './styles';
export * from './renderHelpers';
export * from './AppComponent';
