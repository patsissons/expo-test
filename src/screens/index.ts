export * from './MainScreen';
export * from './LoginScreen';
