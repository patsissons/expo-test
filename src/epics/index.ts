import 'rxjs';
import { Action } from 'redux';
import { combineEpics, Epic } from 'redux-observable';
import { Store } from '../components';
import { navigate } from './navigation';

export const epics: Epic<Action, Store> = combineEpics(
  navigate,
);
