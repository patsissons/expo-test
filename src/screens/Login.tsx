import React from 'react';
import { View, Text } from 'react-native';
import { LoginState, AppComponent, withState } from '../framework';

class LoginScreen extends AppComponent<Login.ScreenProps, LoginState> {
  // istanbul ignore next navigation options
  public static navigationOptions(params: AppComponent.ContextNavigationScreenProps) {
    return {
      title: params.screenProps.context.localization.LoginScreen,
    };
  }

  render() {
    return (
      <View style={ this.S.common.container }>
        <Text>{ this.L.LoginScreen }</Text>
      </View>
    );
  }
}

export namespace Login {
  export interface Props {
  }

  export interface ScreenProps extends AppComponent.Props<LoginState>, Props {
  }
}

export const Login = withState(
  LoginScreen,
  x => x.login,
);
