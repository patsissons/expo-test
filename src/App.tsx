import React from 'react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import { createEpicMiddleware } from 'redux-observable';
import { composeWithDevTools } from 'redux-devtools-extension'
import '../types';
import { reducers } from './reducers';
import { epics } from './epics';
import { AppNavigator, reactNavigationMiddleware } from './navigation/AppNavigator';

const store = createStore(
  reducers,
  composeWithDevTools(
    applyMiddleware(
      createEpicMiddleware(epics as any),
      reactNavigationMiddleware,
    ),
  ),
);

export class App extends React.Component {
  render() {
    return (
      <Provider store={ store }>
        <AppNavigator />
      </Provider>
    );
  }
}
