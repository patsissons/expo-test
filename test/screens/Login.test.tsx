import React from 'react';
import { shallow } from 'enzyme';
import { createStoreProps } from '../storeUtils';
import { Login } from '../../src/screens/Login';
import { Store } from '../../src/framework';

const defaultState: Partial<Store> = {
  login: {},
};

describe('screens', () => {
  describe('Main', () => {
    it.only('renders without crashing', () => {
      // Arrange
      const props = createStoreProps(defaultState);
      const component = (<Login { ...props as any } />);

      // Act
      const result = shallow(component).dive();

      // Assert
      expect(result).toMatchSnapshot();
    });
  });
});
