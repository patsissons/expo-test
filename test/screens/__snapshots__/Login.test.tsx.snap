// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screens Main renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoginScreen
    actions={
      Object {
        "context": Object {
          "loadEnv": [Function],
          "loadLocale": [Function],
        },
        "demo": Object {
          "decrement": [Function],
          "increment": [Function],
        },
        "navigation": Object {
          "goBack": [Function],
          "toLogin": [Function],
          "toMain": [Function],
        },
      }
    }
    context={
      Object {
        "env": Object {
          "RELEASE": "local",
        },
        "locale": "en",
        "localization": Object {
          "CurrentStateIs": "Current state is",
          "Increment": "Increment",
          "IncrementCount": "Increment Count",
          "Login": "Login",
          "LoginScreen": "Login Screen",
          "MainScreen": "Main Screen",
        },
      }
    }
    state={Object {}}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Login Screen
      </Text>,
      "style": Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flex": 1,
        "justifyContent": "center",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "accessible": true,
        "allowFontScaling": true,
        "children": "Login Screen",
        "ellipsizeMode": "tail",
      },
      "ref": null,
      "rendered": "Login Screen",
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          Login Screen
        </Text>,
        "style": Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "center",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "accessible": true,
          "allowFontScaling": true,
          "children": "Login Screen",
          "ellipsizeMode": "tail",
        },
        "ref": null,
        "rendered": "Login Screen",
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
